### ISSUE TITLE
-g or --gitignore doesn't seem to work

### ISSUE THREAD

**Issue by micisse** [ouch-org/ouch#506]
### Version

ouch 0.4.2

### Description

-g or --gitignore doesn't seem to work (see attached video). Or am I doing it wrong? The use of **--gitignore** or **-g** is not mentioned anywhere in the documentation. I found it with **--help** then https://github.com/ouch-org/ouch/issues/170 to https://github.com/ouch-org/ouch/pull/245

The contents of the **.gitignore** file are as follows :

`*.zip`
`*.tar.gz`

https://github.com/ouch-org/ouch/assets/56940294/e0cddc56-f27c-4a29-9b95-d3fd9e544ef5

### Expected Behavior

It should normally ignore files/folders mentioned in **.gitignore**

**Comment by marcospb19** [comment]
This can be a bit confusing, here's how `-g/--gitignore` works:

When recursing through directory entries, do not include the ones in `.gitignore`.

(This can be useful if you want to compress a repository without including the build artifacts.)

However, you can bypass `-g/--gitignore` by providing the explicit name of the files you want.

That's what you're doing, your shell (bash, zsh or fish) expands `*` to the two arguments "PNG.zip" and "SVG.zip", so you're bypassing the `-g` flag.

---

The current workaround would be to pass the folder itself to Ouch.

```sh
cd ..
ouch compress folder testo.tar.gz --gitignore
```

But we can also discuss wether we want to remove the capability to bypass `--gitignore`.

---

```
  -g, --gitignore
          Ignores files matched by git's ignore files
```

Also, we need to update the `--help` to make this clear, and maybe document it in README.md.

**Comment by micisse** [comment]
@marcospb19 Thanks for your response

> However, you can bypass -g/--gitignore by providing the explicit name of the files you want.

The files contained in folders are too numerous to be defined by hand. That's why I took `PNG.zip` and `SVG.zip` to create this issue. 🔥.

> The current workaround would be to pass the folder itself to Ouch.

It wasn't the folder I wanted to compress, but the contents only, and above all to be able to access the files directly when opening the archive (_to avoid **--strip-components=1** with tar for example_).

> That's what you're doing, your shell (bash, zsh or fish) expands * to the two arguments "PNG.zip" and "SVG.zip", so you're bypassing the -g flag.

The idea is to take everything (**\***) in the folder but exclude the files mentioned in **.gitignore** and/or **.gitignore** file itself. To use **git** with **.gitignore** file, you have to be in the folder itself at the root of the **.git** folder and the **git** command takes **.gitignore** into account in the same place, no need to pass the parent folder.

---

Ouch could perhaps be inspired by the "tar" command with --exclude arg. Take inspiration from it, passing several exclusion arguments in the same way (**--exclude "<regex/*.ext/string...etc>"**) otherwise the **.gitignore** file idea is also good to avoid multiple --exclude arguments. 

```sh
tar -czvf testo.tar.gz --exclude "*.zip" --exclude "*.tar.gz" -C <path> .(dot) --overwrite
tar -czvf testo.tar.gz --exclude "*.zip" --exclude "*.tar.gz" .(dot) --overwrite
```

![Capture d’écran_2023-09-11_20-49-46](https://github.com/ouch-org/ouch/assets/56940294/c3373ca1-3689-4ed7-bfc8-eca938985a0d)

The output of "**ouch --help**" gives "**Usage: ouch [OPTIONS] \<COMMAND\>**" the options before the command. Is it necessary to respect the order?

**Comment by micisse** [comment]
@marcospb19 Thanks for your response

> However, you can bypass -g/--gitignore by providing the explicit name of the files you want.

The files contained in folders are too numerous to be defined by hand. That's why I took `PNG.zip` and `SVG.zip` to create this issue. 🔥.

> The current workaround would be to pass the folder itself to Ouch.

It wasn't the folder I wanted to compress, but the contents only, and above all to be able to access the files directly when opening the archive (_to avoid **--strip-components=1** with tar for example_).

> That's what you're doing, your shell (bash, zsh or fish) expands * to the two arguments "PNG.zip" and "SVG.zip", so you're bypassing the -g flag.

The idea is to take everything (**\***) in the folder but exclude the files mentioned in **.gitignore** and/or **.gitignore** file itself. To use **git** with **.gitignore** file, you have to be in the folder itself at the root of the **.git** folder and the **git** command takes **.gitignore** into account in the same place, no need to pass the parent folder.

---

Ouch could perhaps be inspired by the "tar" command with --exclude arg. Passing several exclusion arguments in the same way (**--exclude "<regex/*.ext/string...etc>"**) otherwise the **.gitignore** or **.ouchignore** file idea is also good to avoid multiple --exclude arguments. 

```sh
tar -czvf testo.tar.gz --exclude "*.zip" --exclude "*.tar.gz" -C <path> .(dot) --overwrite
tar -czvf testo.tar.gz --exclude "*.zip" --exclude "*.tar.gz" .(dot) --overwrite
```

![Capture d’écran_2023-09-11_20-49-46](https://github.com/ouch-org/ouch/assets/56940294/c3373ca1-3689-4ed7-bfc8-eca938985a0d)

The output of "**ouch --help**" gives "**Usage: ouch [OPTIONS] \<COMMAND\>**" the options before the command. Is it necessary to respect the order?

**Comment by marcospb19** [comment]
> The output of "**ouch --help**" gives "**Usage: ouch [OPTIONS] <COMMAND>**" the options before the command. Is it necessary to respect the order?

No, `--gitignore` is considered a global flag because it applies to all 3 subcommand.

Thanks for your input, having a `--exclude` flag sounds like a great path to solve this one.

**Comment by the-ge** [comment]
I am not able to make it comply to .gitignore, regardless if .gitignore is inside or outside the folder to be compressed (ouch 0.5.1 on Manjaro KDE) or the flag position in the command. I feel stupid now, what am I missing?

https://github.com/user-attachments/assets/51ae21f0-db4d-4788-9932-0a1e19507815

**Comment by talis-fb** [comment]
@the-ge Is either `test/` or `folder/`  a Git repository? Today the `--gitignore` flag only excludes files that are tracked by Git, so it respects .gitignore only when inside a Git repository.

**Comment by the-ge** [comment]
Ah, thank you, so that's why - those folders are not actually tracked, just attempted to use .gitignore as a sort of exclude file. Sorry about that.

**Comment by marcospb19** [comment]
No problem! Maybe that was @micisse's problem too, now I realize in the screenshot there is no `.git/` folder.

I think it's reasonable to keep the current behavior 🤔. Closing for now.
